opentreemap/importer/models/trees.py

```diff
@@ -250,30 +250,59 @@ def validate_otm_id(self):
         return True
 
     def validate_proximity(self, point):
-        plot_ids_from_this_import = TreeImportRow.objects\
-            .filter(import_event=self.import_event)\
-            .filter(plot__isnull=False)\
-            .values_list('plot__pk', flat=True)
+        """Return False if the point is within 10ft of an existing
+        plot, excluding plots that are part of the same import event.
 
-        nearby = Plot.objects\
-                     .filter(instance=self.import_event.instance)\
-                     .filter(geom__distance_lte=(point, D(ft=10.0)))\
-                     .distance(point)\
-                     .exclude(pk__in=plot_ids_from_this_import)\
+        This is written with as a raw SQL query to take advantage of
+        the PostGIS <-> operator which provides indexed nearest
+        neighbor searching.
 
-        oid = self.cleaned.get(fields.trees.OPENTREEMAP_PLOT_ID, None)
-        if oid:
-            nearby = nearby.exclude(pk=oid)
+        The `AND NOT IN` clause skips plots created by the same import
+        event, assuming that your import file is correct, even though
+        trees are close together.
 
-        nearby = nearby.order_by('distance')[:5]
-
-        if len(nearby) > 0:
+        `LIMIT 2` because we may be updating an existing plot, in
+        which case the nearest plot will be the plot itself, and we
+        will skip over it looking for the next nearest plot.
+        """
+        sql = ('SELECT id, '
+               '  ST_Distance(the_geom_webmercator, '
+               '              ST_SetSRID(ST_MakePoint(%(x)s, %(y)s), '
+               '                         %(srid)s)) '
+               '  AS distance '
+               'FROM treemap_mapfeature '
+               'WHERE instance_id = %(instance_id)s '
+               'AND NOT id IN ( '
+               '  SELECT plot_id '
+               '  FROM importer_treeimportrow tir '
+               '  WHERE tir.import_event_id = %(event_id)s '
+               '  AND tir.plot_id IS NOT NULL) '
+               'ORDER BY the_geom_webmercator <-> '
+               '  ST_SetSRID(ST_MakePoint(%(x)s, %(y)s), %(srid)s) '
+               'LIMIT 2;')
```

Comment:
> `LIMIT 2` because we may be updating an existing plot, in which case the nearest plot will be the plot itself

If we filter out any possible existing plot in the SQL query, can we remove the ``LIMIT 2`?

Maybe by adding something like this to the `WHERE` clause (with `existing_plot_id` possibly being null)?
`AND id IS DISTINCT FROM %(existing_plot_id)`